File: variants/scraper/rentmasseur_scraper.py
"""
RentMasseur.com Bio Watcher and Collector
Automates collection of user bios, view tracking, and views per day analysis
"""

import asyncio
import json
import csv
import re
from datetime import datetime, timedelta
from pathlib import Path
from typing import Dict, List, Optional
from dataclasses import dataclass, asdict
from playwright.async_api import async_playwright, Browser, Page


@dataclass
class MasseurProfile:
    """Data model for masseur profile"""
    username: str
    profile_url: str
    location: str
    registration_date: Optional[str] = None
    total_views: Optional[int] = None
    bio: Optional[str] = None
    massage_types: List[str] = None
    travel_schedule: List[str] = None
    last_updated: str = None
    views_per_day: Optional[float] = None
    
    def __post_init__(self):
        if self.massage_types is None:
            self.massage_types = []
        if self.travel_schedule is None:
            self.travel_schedule = []
        self.last_updated = datetime.now().isoformat()


class RentMasseurScraper:
    """Main scraper class for RentMasseur.com"""
    
    BASE_URL = "https://rentmasseur.com"
    CITY_URL_PATTERN = f"{BASE_URL}/gay-massage/{{city}}"
    PROFILE_URL_PATTERN = f"{BASE_URL}/{{username}}"
    
    def __init__(self, data_dir: str = "data"):
        self.data_dir = Path(data_dir)
        self.data_dir.mkdir(exist_ok=True)
        self.ua = UserAgent()
        self.profiles: Dict[str, MasseurProfile] = {}
        self.session: Optional[aiohttp.ClientSession] = None
        
    async def __aenter__(self):
        self.session = aiohttp.ClientSession()
        return self
    
    async def __aexit__(self, exc_type, exc_val, exc_tb):
        if self.session:
            await self.session.close()
    
    def get_cities(self) -> List[str]:
        """List of cities to scrape"""
        return [
            "newyork", "losangeles", "manhattan-ny", "atlanta", "miami",
            "london", "west-hollywood-ca", "palmsprings", "chicago", "dallas",
            "sanfrancisco", "ftlauderdale", "sandiego", "houston", "lasvegas-nv",
            "toronto", "washingtondc", "orangecounty-ca", "orlando", "philadelphia"
        ]
    
    async def get_profile_links_from_city(self, browser: Browser, city: str) -> List[str]:
        """Extract profile links from a city page"""
        url = self.CITY_URL_PATTERN.format(city=city)
        print(f"Scraping city: {city}")
        
        page = await browser.new_page()
        try:
            await page.goto(url, wait_until="networkidle", timeout=30000)
            await page.wait_for_timeout(2000)  # Wait for dynamic content
            
            # Extract all profile links
            profile_links = await page.eval_on_selector_all(
                'a[href^="/"]',
                'els => els.map(el => el.href).filter(href => href.match(/rentmasseur\\.com\\/[^\\/]+$/))'
            )
            
            # Clean and deduplicate links
            unique_links = list(set([
                link for link in profile_links 
                if link and not link.endswith('/') and len(link.split('/')[-1]) > 2
            ]))
            
            print(f"Found {len(unique_links)} profiles in {city}")
            return unique_links
            
        except Exception as e:
            print(f"Error scraping {city}: {e}")
            return []
        finally:
            await page.close()
    
    async def scrape_profile(self, browser: Browser, profile_url: str) -> Optional[MasseurProfile]:
        """Scrape individual profile data"""
        username = profile_url.rstrip('/').split('/')[-1]
        print(f"Scraping profile: {username}")
        
        page = await browser.new_page()
        try:
            await page.goto(profile_url, wait_until="networkidle", timeout=30000)
            await page.wait_for_timeout(2000)
            
            # Extract profile data
            profile = MasseurProfile(
                username=username,
                profile_url=profile_url
            )
            
            # Location
            try:
                location = await page.text_content('a[href*="gay-massage"]')
                if location:
                    profile.location = location.strip()
            except:
                pass
            
            # Bio/description
            try:
                bio = await page.text_content('.bio, .description, [class*="bio"], [class*="about"]')
                if bio:
                    profile.bio = bio.strip()
            except:
                pass
            
            # Massage types
            try:
                massage_types = await page.eval_on_selector_all(
                    'a[href*="massages/"]',
                    'els => els.map(el => el.textContent.trim())'
                )
                if massage_types:
                    profile.massage_types = massage_types
            except:
                pass
            
            # Travel schedule
            try:
                travel = await page.eval_on_selector_all(
                    '[class*="travel"], [class*="schedule"] a',
                    'els => els.map(el => el.textContent.trim())'
                )
                if travel:
                    profile.travel_schedule = travel
            except:
                pass
            
            # Views (this may require login or be in a specific element)
            try:
                views_text = await page.text_content('[class*="view"], [class*="View"]')
                if views_text:
                    views_match = re.search(r'(\d+[,\d]*)', views_text)
                    if views_match:
                        profile.total_views = int(views_match.group(1).replace(',', ''))
            except:
                pass
            
            # Registration date
            try:
                reg_text = await page.text_content('[class*="member"], [class*="since"], [class*="joined"]')
                if reg_text:
                    date_match = re.search(r'(?:since|joined|member)\s*(.+)', reg_text, re.IGNORECASE)
                    if date_match:
                        profile.registration_date = date_match.group(1).strip()
            except:
                pass
            
            # Calculate views per day if we have both views and registration date
            if profile.total_views and profile.registration_date:
                profile.views_per_day = self.calculate_views_per_day(
                    profile.total_views, 
                    profile.registration_date
                )
            
            return profile
            
        except Exception as e:
            print(f"Error scraping profile {username}: {e}")
            return None
        finally:
            await page.close()
    
    def calculate_views_per_day(self, total_views: int, registration_date: str) -> float:
        """Calculate views per day based on registration date"""
        try:
            # Parse various date formats
            for fmt in ['%Y-%m-%d', '%m/%d/%Y', '%B %Y', '%b %Y', '%Y']:
                try:
                    reg_date = datetime.strptime(registration_date.split()[0], fmt)
                    days_active = (datetime.now() - reg_date).days
                    if days_active > 0:
                        return round(total_views / days_active, 2)
                except:
                    continue
            return None
        except:
            return None
    
    async def scrape_all_profiles(self, headless: bool = True):
        """Main method to scrape all profiles from all cities"""
        async with async_playwright() as p:
            browser = await p.chromium.launch(headless=headless)
            
            all_profile_urls = set()
            
            # Collect all profile URLs from all cities
            for city in self.get_cities():
                try:
                    profile_links = await self.get_profile_links_from_city(browser, city)
                    all_profile_urls.update(profile_links)
                except Exception as e:
                    print(f"Error processing city {city}: {e}")
                    continue
            
            print(f"Total unique profiles found: {len(all_profile_urls)}")
            
            # Scrape each profile
            for url in all_profile_urls:
                try:
                    profile = await self.scrape_profile(browser, url)
                    if profile:
                        self.profiles[profile.username] = profile
                except Exception as e:
                    print(f"Error processing profile {url}: {e}")
                    continue
            
            await browser.close()
    
    def save_to_json(self, filename: str = "masseur_profiles.json"):
        """Save profiles to JSON file"""
        filepath = self.data_dir / filename
        data = {
            username: asdict(profile) 
            for username, profile in self.profiles.items()
        }
        with open(filepath, 'w', encoding='utf-8') as f:
            json.dump(data, f, indent=2, ensure_ascii=False)
        print(f"Saved {len(self.profiles)} profiles to {filepath}")
    
    def save_to_csv(self, filename: str = "masseur_profiles.csv"):
        """Save profiles to CSV file"""
        filepath = self.data_dir / filename
        rows = []
        for profile in self.profiles.values():
            row = asdict(profile)
            row['massage_types'] = ', '.join(profile.massage_types)
            row['travel_schedule'] = ', '.join(profile.travel_schedule)
            rows.append(row)
        
        if rows:
            with open(filepath, 'w', newline='', encoding='utf-8') as f:
                fieldnames = list(rows[0].keys())
                writer = csv.DictWriter(f, fieldnames=fieldnames)
                writer.writeheader()
                writer.writerows(rows)
            print(f"Saved {len(rows)} profiles to {filepath}")
    
    def generate_report(self, filename: str = "views_report.md"):
        """Generate views per day analysis report"""
        filepath = self.data_dir / filename
        
        # Sort by views per day
        sorted_profiles = sorted(
            self.profiles.values(),
            key=lambda x: x.views_per_day or 0,
            reverse=True
        )
        
        with open(filepath, 'w', encoding='utf-8') as f:
            f.write("# RentMasseur Views Analysis Report\n\n")
            f.write(f"Generated: {datetime.now().isoformat()}\n")
            f.write(f"Total Profiles: {len(self.profiles)}\n\n")
            
            f.write("## Top Profiles by Views Per Day\n\n")
            f.write("| Username | Location | Total Views | Views/Day | Registered |\n")
            f.write("|----------|----------|-------------|-----------|------------|\n")
            
            for profile in sorted_profiles[:50]:  # Top 50
                f.write(
                    f"| {profile.username} | {profile.location} | "
                    f"{profile.total_views or 'N/A'} | "
                    f"{profile.views_per_day or 'N/A'} | "
                    f"{profile.registration_date or 'N/A'} |\n"
                )
            
            f.write("\n## All Profiles\n\n")
            for profile in sorted_profiles:
                f.write(f"### {profile.username}\n")
                f.write(f"- **Location**: {profile.location}\n")
                f.write(f"- **Total Views**: {profile.total_views or 'N/A'}\n")
                f.write(f"- **Views/Day**: {profile.views_per_day or 'N/A'}\n")
                f.write(f"- **Registered**: {profile.registration_date or 'N/A'}\n")
                f.write(f"- **Massage Types**: {', '.join(profile.massage_types)}\n")
                f.write(f"- **URL**: {profile.profile_url}\n\n")
        
        print(f"Generated report: {filepath}")


async def main():
    """Main entry point"""
    scraper = RentMasseurScraper()
    
    print("Starting RentMasseur scraper...")
    await scraper.scrape_all_profiles(headless=False)  # Set to True for production
    
    print("\nSaving data...")
    scraper.save_to_json()
    scraper.save_to_csv()
    scraper.generate_report()
    
    print(f"\nScraping complete! Total profiles collected: {len(scraper.profiles)}")


if __name__ == "__main__":
    asyncio.run(main())

File: variants/scraper/rentmasseur_scraper.cpp
#include <iostream>
#include <string>
#include <vector>
#include <map>
#include <set>
#include <fstream>
#include <sstream>
#include <regex>
#include <ctime>
#include <cmath>
#include <thread>
#include <chrono>
#include <algorithm>
#include <curl/curl.h>

struct MasseurProfile {
    std::string username;
    std::string profile_url;
    std::string location;
    std::string registration_date;
    int total_views;
    std::string bio;
    std::vector<std::string> massage_types;
    std::vector<std::string> travel_schedule;
    double views_per_day;
    std::string last_updated;
    
    MasseurProfile() : total_views(0), views_per_day(0.0) {}
};

size_t WriteCallback(void* contents, size_t size, size_t nmemb, void* userp) {
    ((std::string*)userp)->append((char*)contents, size * nmemb);
    return size * nmemb;
}

size_t HeaderCallback(void* contents, size_t size, size_t nmemb, void* userp) {
    std::string* header_string = (std::string*)userp;
    header_string->append((char*)contents, size * nmemb);
    return size * nmemb;
}

std::string cookie_file = "cookies.txt";

std::string login(const std::string& username, const std::string& password) {
    CURL* curl;
    CURLcode res;
    std::string readBuffer;
    
    curl = curl_easy_init();
    if(curl) {
        // Login URL
        std::string login_url = "https://rentmasseur.com/login";
        
        // Post data
        std::string post_fields = "username=" + username + "&password=" + password;
        
        curl_easy_setopt(curl, CURLOPT_URL, login_url.c_str());
        curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, WriteCallback);
        curl_easy_setopt(curl, CURLOPT_WRITEDATA, &readBuffer);
        curl_easy_setopt(curl, CURLOPT_POSTFIELDS, post_fields.c_str());
        curl_easy_setopt(curl, CURLOPT_COOKIEJAR, cookie_file.c_str());
        curl_easy_setopt(curl, CURLOPT_COOKIEFILE, cookie_file.c_str());
        curl_easy_setopt(curl, CURLOPT_FOLLOWLOCATION, 1L);
        curl_easy_setopt(curl, CURLOPT_USERAGENT, "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36");
        curl_easy_setopt(curl, CURLOPT_TIMEOUT, 30L);
        
        res = curl_easy_perform(curl);
        if(res != CURLE_OK) {
            std::cerr << "Login failed: " << curl_easy_strerror(res) << std::endl;
        } else {
            std::cout << "Login attempt completed" << std::endl;
        }
        curl_easy_cleanup(curl);
    }
    return readBuffer;
}

std::string fetch_url(const std::string& url, bool use_cookies = true) {
    CURL* curl;
    CURLcode res;
    std::string readBuffer;
    
    curl = curl_easy_init();
    if(curl) {
        curl_easy_setopt(curl, CURLOPT_URL, url.c_str());
        curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, WriteCallback);
        curl_easy_setopt(curl, CURLOPT_WRITEDATA, &readBuffer);
        curl_easy_setopt(curl, CURLOPT_FOLLOWLOCATION, 1L);
        curl_easy_setopt(curl, CURLOPT_USERAGENT, "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36");
        curl_easy_setopt(curl, CURLOPT_TIMEOUT, 30L);
        
        if(use_cookies) {
            curl_easy_setopt(curl, CURLOPT_COOKIEFILE, cookie_file.c_str());
            curl_easy_setopt(curl, CURLOPT_COOKIEJAR, cookie_file.c_str());
        }
        
        res = curl_easy_perform(curl);
        if(res != CURLE_OK) {
            std::cerr << "curl_easy_perform() failed: " << curl_easy_strerror(res) << std::endl;
        }
        curl_easy_cleanup(curl);
    }
    return readBuffer;
}

std::vector<std::string> extract_profile_links(const std::string& html) {
    std::vector<std::string> links;
    std::regex profile_regex(R"(https://rentmasseur\.com/([a-zA-Z0-9_-]+))");
    std::sregex_iterator it(html.begin(), html.end(), profile_regex);
    std::sregex_iterator end;
    
    std::set<std::string> unique_links;
    for(; it != end; ++it) {
        std::string link = it->str();
        if(link.find("gay-massage") == std::string::npos && 
           link.find("massages/") == std::string::npos &&
           link.length() > 25) {
            unique_links.insert(link);
        }
    }
    
    for(const auto& link : unique_links) {
        links.push_back(link);
    }
    
    return links;
}

std::string extract_text(const std::string& html, const std::string& pattern) {
    std::regex regex(pattern);
    std::smatch match;
    if(std::regex_search(html, match, regex) && match.size() > 1) {
        return match[1].str();
    }
    return "";
}

int extract_views(const std::string& html) {
    std::regex view_regex(R"((\d{1,3}(,\d{3})*))");
    std::sregex_iterator it(html.begin(), html.end(), view_regex);
    std::sregex_iterator end;
    
    int max_views = 0;
    for(; it != end; ++it) {
        std::string num_str = (*it)[1].str();
        // Remove commas
        num_str.erase(std::remove(num_str.begin(), num_str.end(), ','), num_str.end());
        try {
            int num = std::stoi(num_str);
            if(num > max_views && num < 1000000) { // Reasonable view count range
                max_views = num;
            }
        } catch(...) {}
    }
    return max_views;
}

std::string extract_registration_date(const std::string& html) {
    std::vector<std::string> patterns = {
        R"(since\s+([A-Za-z]+\s+\d{4}))",
        R"(joined\s+([A-Za-z]+\s+\d{4}))",
        R"(member\s+since\s+([A-Za-z]+\s+\d{4}))",
        R"((\d{4}-\d{2}-\d{2}))",
        R"((\d{2}/\d{2}/\d{4}))"
    };
    
    for(const auto& pattern : patterns) {
        std::string result = extract_text(html, pattern);
        if(!result.empty()) {
            return result;
        }
    }
    return "";
}

double calculate_views_per_day(int total_views, const std::string& reg_date_str) {
    if(total_views == 0 || reg_date_str.empty()) return 0.0;
    
    // Try to parse the date
    std::tm tm = {};
    std::istringstream iss(reg_date_str);
    
    // Try different formats
    const char* formats[] = {"%Y-%m-%d", "%m/%d/%Y", "%B %Y", "%b %Y", "%Y"};
    for(const auto& fmt : formats) {
        iss.clear();
        iss.str(reg_date_str);
        if(iss >> std::get_time(&tm, fmt)) {
            std::time_t reg_time = std::mktime(&tm);
            std::time_t now = std::time(nullptr);
            double days = difftime(now, reg_time) / (60 * 60 * 24);
            if(days > 0) {
                return round((total_views / days) * 100.0) / 100.0;
            }
        }
    }
    
    return 0.0;
}

std::string get_current_timestamp() {
    std::time_t now = std::time(nullptr);
    char buf[80];
    std::strftime(buf, sizeof(buf), "%Y-%m-%dT%H:%M:%S", std::localtime(&now));
    return std::string(buf);
}

void save_to_json(const std::map<std::string, MasseurProfile>& profiles, const std::string& filename) {
    std::ofstream file(filename);
    file << "{\n";
    
    auto it = profiles.begin();
    while(it != profiles.end()) {
        const auto& profile = it->second;
        file << "  \"" << profile.username << "\": {\n";
        file << "    \"username\": \"" << profile.username << "\",\n";
        file << "    \"profile_url\": \"" << profile.profile_url << "\",\n";
        file << "    \"location\": \"" << profile.location << "\",\n";
        file << "    \"registration_date\": \"" << profile.registration_date << "\",\n";
        file << "    \"total_views\": " << profile.total_views << ",\n";
        file << "    \"views_per_day\": " << profile.views_per_day << ",\n";
        file << "    \"bio\": \"" << profile.bio << "\",\n";
        file << "    \"last_updated\": \"" << profile.last_updated << "\"\n";
        file << "  }";
        
        if(++it != profiles.end()) {
            file << ",\n";
        } else {
            file << "\n";
        }
    }
    
    file << "}\n";
    file.close();
    std::cout << "Saved " << profiles.size() << " profiles to " << filename << std::endl;
}

void save_to_csv(const std::map<std::string, MasseurProfile>& profiles, const std::string& filename) {
    std::ofstream file(filename);
    file << "username,profile_url,location,registration_date,total_views,views_per_day,bio,last_updated\n";
    
    for(const auto& pair : profiles) {
        const auto& profile = pair.second;
        file << "\"" << profile.username << "\","
             << "\"" << profile.profile_url << "\","
             << "\"" << profile.location << "\","
             << "\"" << profile.registration_date << "\","
             << profile.total_views << ","
             << profile.views_per_day << ","
             << "\"" << profile.bio << "\","
             << "\"" << profile.last_updated << "\"\n";
    }
    
    file.close();
    std::cout << "Saved " << profiles.size() << " profiles to " << filename << std::endl;
}

void generate_report(const std::map<std::string, MasseurProfile>& profiles, const std::string& filename) {
    std::ofstream file(filename);
    
    file << "# RentMasseur Views Analysis Report\n\n";
    file << "Generated: " << get_current_timestamp() << "\n";
    file << "Total Profiles: " << profiles.size() << "\n\n";
    
    // Sort by views per day
    std::vector<MasseurProfile> sorted_profiles;
    for(const auto& pair : profiles) {
        sorted_profiles.push_back(pair.second);
    }
    std::sort(sorted_profiles.begin(), sorted_profiles.end(), 
              [](const MasseurProfile& a, const MasseurProfile& b) {
                  return a.views_per_day > b.views_per_day;
              });
    
    file << "## Top Profiles by Views Per Day\n\n";
    file << "| Username | Location | Total Views | Views/Day | Registered |\n";
    file << "|----------|----------|-------------|-----------|------------|\n";
    
    size_t count = 0;
    for(const auto& profile : sorted_profiles) {
        if(count++ >= 50) break;
        file << "| " << profile.username << " | " << profile.location << " | "
             << profile.total_views << " | " << profile.views_per_day << " | "
             << profile.registration_date << " |\n";
    }
    
    file.close();
    std::cout << "Generated report: " << filename << std::endl;
}

int main() {
    curl_global_init(CURL_GLOBAL_DEFAULT);
    
    // First, fetch login page to see form structure
    std::cout << "Fetching login page..." << std::endl;
    std::string login_page = fetch_url("https://rentmasseur.com/login", false);
    std::ofstream login_debug("login_page.html");
    login_debug << login_page;
    login_debug.close();
    std::cout << "Login page saved to login_page.html" << std::endl;
    
    // Login first
    std::cout << "Logging in as karpathianwolf..." << std::endl;
    std::string login_response = login("karpathianwolf", "Lola369!");
    
    std::ofstream login_response_debug("login_response.html");
    login_response_debug << login_response;
    login_response_debug.close();
    std::cout << "Login response saved to login_response.html" << std::endl;
    
    std::map<std::string, MasseurProfile> profiles;
    
    std::vector<std::string> cities = {
        "newyork", "losangeles", "manhattan-ny", "atlanta", "miami",
        "london", "west-hollywood-ca", "palmsprings", "chicago", "dallas",
        "sanfrancisco", "ftlauderdale", "sandiego", "houston", "lasvegas-nv",
        "toronto", "washingtondc", "orangecounty-ca", "orlando", "philadelphia"
    };
    
    std::set<std::string> all_profile_urls;
    
    // First, try scraping the homepage for featured profiles
    std::string homepage_url = "https://rentmasseur.com/";
    std::cout << "Scraping homepage for featured profiles..." << std::endl;
    
    std::string homepage_html = fetch_url(homepage_url);
    if(!homepage_html.empty()) {
        std::ofstream debug_file("debug_homepage.html");
        debug_file << homepage_html;
        debug_file.close();
        
        std::cout << "Downloaded " << homepage_html.length() << " bytes from homepage" << std::endl;
        
        std::vector<std::string> homepage_links = extract_profile_links(homepage_html);
        for(const auto& link : homepage_links) {
            all_profile_urls.insert(link);
        }
        
        std::cout << "Found " << homepage_links.size() << " profiles on homepage" << std::endl;
    }
    
    // Also add known profiles from the homepage we saw earlier
    std::vector<std::string> known_profiles = {
        "https://rentmasseur.com/Ritual",
        "https://rentmasseur.com/JonnasLatino",
        "https://rentmasseur.com/LiamGoodBoy",
        "https://rentmasseur.com/BrunoMathias",
        "https://rentmasseur.com/HOLLYHOODONLYGEN",
        "https://rentmasseur.com/YULIAN",
        "https://rentmasseur.com/MalikXL",
        "https://rentmasseur.com/BigHandsHK",
        "https://rentmasseur.com/HungMasseurNYC",
        "https://rentmasseur.com/InosukeTopXL",
        "https://rentmasseur.com/Muscltomuscl",
        "https://rentmasseur.com/FemboyFey",
        "https://rentmasseur.com/LVM",
        "https://rentmasseur.com/JayMassive",
        "https://rentmasseur.com/MagicHandsPro",
        "https://rentmasseur.com/ExoticYoungGuy",
        "https://rentmasseur.com/Will_Xavier",
        "https://rentmasseur.com/TonyAsian",
        "https://rentmasseur.com/karpathianwolf"
    };
    
    for(const auto& profile : known_profiles) {
        all_profile_urls.insert(profile);
    }
    
    std::cout << "Total unique profiles to scrape: " << all_profile_urls.size() << std::endl;
    
    // Scrape each profile
    for(const auto& profile_url : all_profile_urls) {
        std::string username = profile_url.substr(profile_url.find_last_of('/') + 1);
        std::cout << "Scraping profile: " << username << std::endl;
        
        std::string html = fetch_url(profile_url);
        if(html.empty()) {
            std::cout << "Failed to fetch " << profile_url << std::endl;
            continue;
        }
        
        MasseurProfile profile;
        profile.username = username;
        profile.profile_url = profile_url;
        profile.location = extract_text(html, R"(Gay Massage in ([^<]+))");
        profile.registration_date = extract_registration_date(html);
        profile.total_views = extract_views(html);
        profile.bio = extract_text(html, R"(<meta name=\"description\" content=\"([^\"]+)\")");
        profile.last_updated = get_current_timestamp();
        profile.views_per_day = calculate_views_per_day(profile.total_views, profile.registration_date);
        
        profiles[username] = profile;
        
        // Small delay to be respectful
        std::this_thread::sleep_for(std::chrono::milliseconds(500));
    }
    
    // Save data
    std::string data_dir = "data";
    system(("mkdir -p " + data_dir).c_str());
    
    save_to_json(profiles, data_dir + "/masseur_profiles.json");
    save_to_csv(profiles, data_dir + "/masseur_profiles.csv");
    generate_report(profiles, data_dir + "/views_report.md");
    
    std::cout << "\nScraping complete! Total profiles collected: " << profiles.size() << std::endl;
    
    curl_global_cleanup();
    return 0;
}
